receiver_test.go

```diff
@@ -2,137 +2,156 @@ package amqp
-	var tag []byte
-	var done chan struct{}
-	if mode == ModeSecond {
-		tag = []byte("one")
-		done = make(chan struct{})
-	}
-	return Message{
-		deliveryID:  uint32(1),
-		DeliveryTag: tag,
-		doneSignal:  done,
-	}
-}
+	"github.com/Azure/go-amqp/internal/encoding"
+	"github.com/Azure/go-amqp/internal/mocks"
+	"github.com/stretchr/testify/assert"
+)
 
-func TestReceiver_HandleMessageModeFirst_AutoAccept(t *testing.T) {
-	r := &Receiver{
-		link:         makeLink(ModeFirst),
-		batching:     true, // allows to  avoid making the outgoing call on dispostion
-		dispositions: make(chan messageDisposition, 2),
-	}
-	msg := makeMessage(ModeFirst)
-	r.link.Messages <- msg
-	if r.link.countUnsettled() != 0 {
-		// mode first messages have no delivery tag, thus there should be no unsettled message
-		t.Fatal("expected zero unsettled count")
+func TestReceive_ModeFirst(t *testing.T) {
+	mock := mocks.NewConnection()
+	// create sequence of AMQP handshake frames
+	mock.AppendProtoHeader(mocks.ProtoAMQP)
+	assert.NoError(t, mock.AppendPerformOpen("test"))
+	client, err := New(mock)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.NoError(t, mock.AppendPerformBegin(0))
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
+	session, err := client.NewSession(ctx)
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	const linkName = "test"
+	const linkHandle = 0
+	assert.NoError(t, mock.AppendReceiverAttach(linkName, linkHandle, ModeFirst))
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	r, err := session.NewReceiver(ctx, LinkName(linkName), LinkReceiverSettle(ModeFirst))
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	// create a mock message for retrieval
+	deliveryID := uint32(1)
+	assert.NoError(t, mock.AppendTransfer(linkHandle, deliveryID, []byte("hello")))
+	// create accepting disposition frame as ModeFirst will automatically accept the message
+	assert.NoError(t, mock.AppendDisposition(deliveryID, &encoding.StateAccepted{}))
+	ctx, cancel = context.WithTimeout(context.Background(), 100*time.Second)
+	msg, err := r.Receive(ctx)
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if c := r.link.countUnsettled(); c != 0 {
+		t.Fatalf("unexpected unsettled count %d", c)
+	}
+	// wait for the link to unpause as credit should now be available
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	for {
+		if atomic.LoadUint32(&r.link.Paused) == 0 {
+			cancel()
+			break
+		} else if err := ctx.Err(); err != nil {
+			t.Fatal(err)
+		}
 	}
-	if err := r.HandleMessage(context.TODO(), doNothing); err != nil {
-		t.Errorf("HandleMessage() error = %v", err)
+	// link credit should be 1
+	if c := r.link.linkCredit; c != 1 {
+		t.Fatalf("unexpected link credit %d", c)
 	}
-
-	if len(r.dispositions) != 0 {
-		t.Errorf("the message should not have triggered a disposition")
+	// subsequent dispositions should have no effect
+	if err := r.AcceptMessage(context.Background(), msg); err != nil {
```

Comment:
It seems like this should actually be an error. I can see no-oping at an application stack above this but it's a constraint violation to attempt to settle a message in mode first (or to settle a message twice, for instance).

Reply:
I couldn't find anything in the spec about this (possible I missed it).  Do you know what Rhea does here?  I can check .NET.

Reply:
I agree. At the amqp level, I would make this an error. If you do settle twice, you'd get an error from the server too.
Would prefer a typed error though, so that we can check on on it safely.

Reply:
This doesn't appear to be the case, i.e. I can accept a message multiple times without error.

Reply:
I somewhat misspoke.  While this works in ModeFirst, in ModeSecond, the second call to AcceptMessage() hangs :(